woocommerce/woocommerce
# [Beta Tester] Add tool to force wccom endpoint errors or timeout

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Closes https://github.com/woocommerce/woocommerce/issues/50494

This PR adds a tool to the Beta Tester to force returned errors on all requests to `woocommerce.com` endpoints to simulate a down or slow server on woocommerce.com.

<img width="788" alt="image" src="https://github.com/user-attachments/assets/9b7d5799-da89-4c7a-a9c1-23390d06206c">



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Its easiest to test this one locally.

1. Tools > WCA Helper > Tools > Force errors on woocommerce.com requests.
2. Choose timeout as the error mode and click "Run".
3. Clear your transients to make sure were not reading from cache `wp transient delete --all`.
4. Go the homescreen and note the slow load time. There are 4 requests to wccom being made on the backend and each will sleep for 6 seconds and result in a timeout error.
5. As of https://github.com/woocommerce/woocommerce/pull/50521 the errors will result in default values being used. Refresh the Homescreen and note normal load times as the cached values are used and a new request is avoided.
6. Now try the Error mode in WCA Test Helper and clear your cache once again `wp transient delete --all`.
7. The homescreen should now load normally, once again with default values as a result of the errors. No errors should be visible on the client.
8. Bonus testing: Add logging to the data poller to confirm errors are being handled. Add some to the Payment Gateway Suggestions as well to see default values being returned as a result.

https://github.com/woocommerce/woocommerce/blob/175ebe2a5e7cc3205ebabc0ef7f10fe02deb69eb/plugins/woocommerce/src/Admin/RemoteSpecs/DataSourcePoller.php#L196

https://github.com/woocommerce/woocommerce/blob/175ebe2a5e7cc3205ebabc0ef7f10fe02deb69eb/plugins/woocommerce/src/Admin/Features/PaymentGatewaySuggestions/Init.php#L74





### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

Add a tool to simulate server problems with woocommerce.com

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
